### PR TITLE
Parse multiple @top statements

### PR DESCRIPTION
**Pull request by nikku**
This implements parsing of multiple `@top` statements in the generator.

Related to https://github.com/lezer-parser/lezer/issues/10

### CODE REVIEW

**Comment by marijnh** [comment]
Yes, this looks good. I guess the next question is how the parser will map from top names to start states—right now, state 0 is hard coded as the start state, but with this, I suppose you'll have to include a map from top rule names to start state numbers in the parser (and the `Parser.deserialize` call). Might keep 0 as the default top rule's start state, so that the parser knows where to start when the `top` option is missing.

(Also, maybe add an error when a top rule that's not the first one doesn't have a name, since you wouldn't be able to address such a rule.)

**Comment by nikku** [comment]
I've added error handling for the unnamed secondary `@top` case. Looks better now.

File: src/grammar.ts [hunk 1]
```diff
@@ -107,10 +109,11 @@ export class TermSet {
     this.terms = this.terms.filter(t => t.terminal || t.preserve || rules.some(r => r.name == t || r.parts.includes(t)))
 
     let names: {[id: number]: string} = {}
-    let nodeTypes = [this.error, this.top]
+    let nodeTypes = [this.error, ...this.tops]
 
     this.error.id = T.Err
-    this.top.id = T.Top
+    for (const term of this.tops) term.id = T.Top
```
Comment (marijnh) [inline src/grammar.ts]: I don't think this'll work—each term needs its own id, so we'll have to stop using a hard-coded top id.

**Comment by marijnh** [comment]
```
    for (const term of this.tops) term.id = T.Top
```

I don't think this'll work—each term needs its own id, so we'll have to stop using a hard-coded top id.

File: src/grammar.ts (hunk 1 above)
Comment (nikku) [reply src/grammar.ts]: Fixed via https://github.com/lezer-parser/lezer-generator/pull/1/commits/1298f5d6659b87665bad111b25f212d7a0938316.

**Comment by nikku** [comment]
So at this point the open issues are:

* `T.Top` cannot be used anymore and should be removed [here](https://github.com/lezer-parser/lezer/blob/master/src/stack.ts#L311)
* `Top -> Start` state has to be serialized and properly de-serialized

For both I'd appreciate your help, as I have no idea how to proceed.

**Comment by marijnh** [comment]
Ah, right, we need the top term's id in the parser as well, when building a tree. So I guess the mapping should be from rule names to `[term, startState]` pairs. Maybe we should, for clarity, get rid of magic state and term ids for the default top rule entirely, and always look those up in the table (when no `top` option is provided, just use `Object.keys(topTable)[0]`).

Building up that table shouldn't be too hard. I guess the parse state builder should return an array of states matching the given start rules. Though preserving those through the state merging probably requires threading that array through `collapseAutomaton` and `mergeIdentical`, which must recompute it when states are merged.

The map then becomes a new argument to `new Parser` and `Parser.deserialize`, which store it in the `Parser` instance and consult it on each parse. I guess the top term id must be stored in the `ParseContext` in order to be able to create the tree at the end of the parse.

Then the `Term.Top` constant can be removed entirely.

**Comment by nikku** [comment]
What you write makes sense to me, however I'm not able to map any of the concepts to the actual source code. As an example, looking at the [`parse.ts`](https://github.com/lezer-parser/lezer/blob/master/src/parse.ts) file I have no idea how the start state to parse from is selected. What do you refer to as magic state?

So yea, I'll stop exploring for now as the remaining issues seem impossible for me to contribute. Maybe what I've started is of use for you or someone else to pic it up.

**Comment by marijnh** [comment]
>  I have no idea how the start state to parse from is selected.

The static function `Stack.start` selects it—it's the 3rd argument to the `Stack` constructor. (I just noticed it was a bit of a mess earlier today and added patch https://github.com/lezer-parser/lezer/commit/fa3270fbd9ee105df6b393745e1ec71b50e038a9, so make sure you have merged that in.)

> So yea, I'll stop exploring for now as the remaining issues seem impossible for me to contribute.

Are you sure? You seem very close to success — assuming the state generation stuff works, it's just a matter of tying things together now.

**Comment by nikku** [comment]
> Are you sure? You seem very close to success — assuming the state generation stuff works, it's just a matter of tying things together now.

Sure. In the end I'm only a poor user, after all. 

If there is test cases missing or anything else not-as-internal. I can help.

**Comment by marijnh** [comment]
Merged as 222f269d8e...67ffa673624a2fc

**Pull request closed without merging**